torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3157,6 +3126,35 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # github.com/NVIDIA/apex/blob/master/apex/parallel/distributed.py
                 orig_grad_data.record_stream(self._streams["post_backward"])
 
+    def _cast_grad_to_param_dtype(
+        self,
+        grad: torch.Tensor,
+        param: FlatParameter,
+    ):
+        """
+        Casts gradient ``grad`` back to the full parameter dtype so that the
+        optimizer step runs with that dtype. This performs an actual cast if
+        1. parameters were in reduced precision during the forward since then
+        gradients would be in that reduced precision, or
+        2. parameters were not in reduced precision but gradients were in
+        reduced precision for communication.
+        However, if a low precision communication hook is registered, then this
```

Comment:
This logic is getting a bit complicated - maybe it would be good to figure out a way to simplify this, such as make the hooks composable and register a low precision comm hook that does the casting if we end up doing mixed precision for gradient communication.